src/services/ProveGatewayService.ts

```diff
@@ -0,0 +1,181 @@
+import BigNumber from 'bignumber.js';
+import { GatewayRepository } from '../repositories/GatewayRepository';
+import { MessageRepository } from '../repositories/MessageRepository';
+import Logger from '../Logger';
+import Utils from '../Utils';
+import { AUXILIARY_GAS_PRICE } from '../Constants';
+import Observer from '../observer/Observer';
+import { AuxiliaryChain } from '../repositories/AuxiliaryChainRepository';
+
+const Mosaic = require('@openst/mosaic.js');
+
+export default class ProveGatewayService extends Observer<AuxiliaryChain> {
+  private gatewayRepository: GatewayRepository;
+
+  private messageRepository: MessageRepository;
+
+  private originWeb3: object;
+
+  private auxiliaryWeb3: object;
+
+  private auxiliaryWorkerAddress: string;
+
+  private gatewayAddress: string;
+
+  private auxiliaryChainId: number;
+
+  /**
+   *  Constructor
+   *
+   * @param gatewayRepository Instance of auxiliary chain repository.
+   * @param messageRepository Instance of message repository.
+   * @param originWeb3 Origin Web3 instance.
+   * @param auxiliaryWeb3 Auxiliary Web3 instance.
+   * @param auxiliaryWorkerAddress auxiliary worker address, this should be
+   *                               unlocked and added in web3 wallet.
+   * @param gatewayAddress Address of gateway contract on origin chain.
+   * @param auxiliaryChainId Auxiliary chain Id.
+   */
+  public constructor(
+    gatewayRepository: GatewayRepository,
+    messageRepository: MessageRepository,
+    originWeb3: object,
+    auxiliaryWeb3: object,
+    auxiliaryWorkerAddress: string,
+    gatewayAddress: string,
+    auxiliaryChainId: number,
+  ) {
+    super();
+
+    this.gatewayRepository = gatewayRepository;
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.auxiliaryWorkerAddress = auxiliaryWorkerAddress;
+    this.gatewayAddress = gatewayAddress;
+    this.messageRepository = messageRepository;
+    this.auxiliaryChainId = auxiliaryChainId;
+
+    // Suppressing lint error
+    // Refer: https://github.com/typescript-eslint/typescript-eslint/issues/636
+    /* eslint-disable @typescript-eslint/unbound-method */
+    this.reactTo = this.reactTo.bind(this);
+    this.update = this.update.bind(this);
+  }
+
+
+  /**
+   * This method react on changes in auxiliary chain models.
+   * @param auxiliaryChains List of auxiliary chains model
+   */
+  public async update(auxiliaryChains: AuxiliaryChain[]): Promise<void> {
+    const proveGatewayPromises = auxiliaryChains
+      .filter(ac => ac.chainId === this.auxiliaryChainId)
+      .map(
+        async (auxiliaryChain): Promise<{
+          success: boolean;
+          transactionHash: string;
+          message: string;
+        }> => (
+          auxiliaryChain.lastOriginBlockHeight
+            ? this.reactTo(auxiliaryChain.lastOriginBlockHeight)
+            : Promise.reject(new Error('Last anchored origin block height cannot be undefined.'))),
```

Comment:
I think this is something which shouldn't happen in the system. This service should always invoke on anchor handler where lastOriginBlockHeight is updated. That's the reason I am rejecting it. 

On the sideline- We should discuss exception handler also. Who will handle exceptions occurred in the repository, handlers and services and how to handle them.

Reply:
That's not true. The service reacts to repository change. If in the repository a record will be created with null lastOriginBlockHeight the service will be triggered. However, service should filter out this case and not react instead of failing (rejecting with error).

Reply:
Done